Core: Add `FixedVector` template - a collection that can be used completely on the stack.

Ivorforce opened the pull request:
This is a high performance `Vector`-like object. In practice, we can use it to reduce the number of per-frame allocations throughout Godot.

In particular, `FixedVector` is useful if the maximum number of objects is small and known, and the objects are needed only temporarily. It could also be used to return a small, uncertain number of objects from a function in a performant fashion if the maximum is known.

The class can be expanded with more useful functions in the future as needed.

## Comparisons
`FixedVector` is designed to be as easily exchangeable for [`Vector`](https://github.com/godotengine/godot/blob/master/core/templates/vector.h) and [`LocalVector`](https://github.com/godotengine/godot/blob/master/core/templates/local_vector.h) as possible. This can be most easily achieved through [`Span`](https://github.com/godotengine/godot/blob/master/core/templates/span.h) interfaces.

`FixedVector` is similar to 3.x' [`FixedArray`](https://github.com/godotengine/godot/blob/3.x/core/fixed_array.h), but some decisions about its design are different.
`FixedVector` is similar to [`std::array`](https://en.cppreference.com/w/cpp/container/array), but it supports adding and removing objects, and a number of other goodies (like conversion to `Span`).
`FixedVector` is somewhat similar to [`StringBuffer`](https://github.com/godotengine/godot/blob/master/core/string/string_buffer.h), but focuses on being a general purpose collection and does not support arbitrary expansion into the heap.

ze2j commented:
Love it! Very useful in many cases (like to store the neighbors of a cell for example).

`StaticVector` is usualy the name given to this kind of fixed capacity container. Why not use that name ?

Ivorforce commented:
> Love it! Very useful in many cases (like to store the neighbors of a cell for example).
> 
> `StaticVector` is usualy the name given to this kind of fixed capacity container. Why not using that name ?

I've seen `Fixed-` used as a prefix for collections where the size is known at compile time — hence I chose the name `FixedVector`. 
I've not seen `Static` used for this concept. Can you quote some examples where this data type is named like this?

ze2j commented:
[Boost static_vector](https://beta.boost.org/doc/libs/1_58_0/doc/html/boost/container/static_vector.html) for example

ze2j commented:
Could you explain what's the rationale for using placement new instead of a relying on a `std::array<T, CAPACITY>` ? I don't get it. Is to pack the data more efficiently ?

Ivorforce commented:
> Could you explain what's the rationale for using placement new instead of a relying on a `std::array<T, CAPACITY>` ? I don't get it. Is to pack the data more efficiently ?

Of course! The reason is performance.

Basically, using `std::array<T, CAPACITY>` will automatically populate the entire array with elements upon construction. For trivial types, like `int`, this is unimportant, and makes no difference.

For non-trivial types though, like `Variant`, this would necessitate the whole array to initialize with an initial value (`NULL`). In addition, pushing new objects requires assigning the new one, which involves branches that account for possible old data. Both waste time.
By using a raw data array, we can skip these computations for improved efficiency.

Repiteo commented:
No preference on the name—both convey the same general idea

lawnjelly commented:
imo `Fixed` is better in the name (as I did in 3.x for `FixedArray`) as the lifetime usually _isn't_ static.

It can either be global/static, or more commonly on the stack. One alternative I suggested in RC was `StackVector`, but for the same reason I wasn't super keen on this, as the lifetime can also be static (although I doubt we'll use this much in Godot).

lawnjelly commented on core/templates/fixed_vector.h:
```diff
@@ -0,0 +1,156 @@
+		} else {
+			for (T &element : p_from) {
+				memnew_placement(ptr() + _size++, T(std::move(element)));
+			}
```
Has the size itself been copied?

lawnjelly commented:
Just to throw an idea, I know people *love* STL, but I go by the philosophy of if a routine in a template (or a header in general) can fail, and it returns `void`, return a `bool` instead for success or failure. This will be compiled out if you don't use it, and provides extra info if you want it.

e.g. `bool push_back(..)`:
```
if (vec.push_back(1))
{
...
}
else
{
// full
}
```
versus
```
if (!vec.is_full())
{
vec.push_back(1);
}
else
{
// full
}
```
I guess the second branch in the `push_back` will be compiled out, but I like the emphasis that this can fail. Same with `pop_back()`.

Ivorforce commented:
> Just to throw an idea, I know people _love_ STL, but I go by the philosophy of if a routine in a template (or a header in general) can fail, and it returns `void`, return a `bool` instead for success or failure. This will be compiled out if you don't use it, and provides extra info if you want it.

This would definitely be a valid way to handle this!
My current thinking was that if you push to a `FixedVector` that's lacking space, you _probably_ have a bug on your hand because you didn't make it large enough statically. So a failure would be most appropriate. But I could also see `FixedVector` being used in scenarios where a push against a full vector is reasonable behavior.

As for `pop_back()`, I thought about it returning the popped `T`, as Rust likes to do. But we don't do this in the codebase right now and it's admittedly less useful than in Rust, which likes to optimize object lifetimes as much as possible. I could also see popping from an empty `FixedVector` returning `bool` (I think I saw that in a library or two as well?), but I think it might more often be a bug we shouldn't ignore, especially in the stack case.

lawnjelly commented:
> As for pop_back(), I also considered returning the popped T, as Rust likes to do.

That's also something I've considered in various forms in the past, although returning a `T` means you have no good options in the case of popping from an empty `vector` (except crashing or returning `T()`).

kiroxas commented on core/templates/fixed_vector.h:
```diff
@@ -0,0 +1,157 @@
+	constexpr void pop_back() {
+		ERR_FAIL_COND(_size == 0);
+		_size--;
+		ptr()[_size].~T();
```
no `if constexpr (!std::is_trivially_destructible_v<T>)` here ?

Ivorforce replied:
I think it just gets compiled away for trivial types:

https://godbolt.org/z/71oWoY19P

I think technically we should expect the same in all the loop cases as well, but the reason I'm checking there is to make it easier for the compiler / not to risk it leaving in the loop for whatever reason.

kiroxas commented on core/templates/fixed_vector.h:
```diff
@@ -0,0 +1,157 @@
+	static_assert(!force_trivial || std::is_trivially_destructible_v<T>, "force_trivial may only be true if the type is trivially destructible.");
+
+	uint32_t _size = 0;
+	alignas(T) uint8_t _data[CAPACITY * sizeof(T)];
```
if you wrap this inside a struct and declare one like : 
```
struct FixedVectorInner
{
     uint32_t _size = 0;
     alignas(T) uint8_t _data[CAPACITY * sizeof(T)];
};
FixedVectorInner inner;
```
Then you don't need the DATA_PADDING information as you can just memcpy with adress of inner and sizeof(inner).  Just a remark.

kiroxas replied:
Well you still can with `memcpySize = &(inner.data + size * sizeof(T)) - &inner` but that's not really better than the current one, agreed.

Ivorforce replied:
That's also an interesting idea! C++ sure can do the same thing in so many different ways 😄
I think for now I still like the data padding best, because it's only awkward in its construction but looks pretty normal elsewhere.

Calinou commented on core/templates/fixed_vector.h:
```diff
@@ -0,0 +1,153 @@
+
+#pragma once
+
+// A high speed Vector of fixed capacity.
+// Especially useful if you need to create an array on the stack, to
+// prevent dynamic allocations (especially in bottleneck code).
+
+// Choose CAPACITY such that it is enough for all elements that could be added through all branches.
+
+// Set force_trivial to true to gain a slight performance improvement on resize calls
+//  if the default value of the type can be ignored.
```
Rework the comment above the class so it displays in full in VS Code when hovered:

```suggestion
/**
 * A high-speed Vector of fixed capacity.
 *
 * Especially useful if you need to create an array on the stack,
 * as it prevents dynamic allocations (especially in bottleneck code).
 *
 * Choose `CAPACITY` such that it is enough for all elements that could be added through all branches.
 * Set `force_trivial` to `true` to gain a slight performance improvement on resize calls
 * if the default value of the type can be ignored.
 */
```

![image](https://github.com/user-attachments/assets/d22b00c8-ced8-4ba9-a325-b026da2eca40)

Ivorforce replied:
Good idea! We don't seem to do this often enough in the codebase as a whole :)

Calinou commented on tests/core/templates/test_fixed_vector.h:
```diff
@@ -0,0 +1,82 @@
+
+namespace TestFixedVector {
+
+TEST_CASE("[FixedVector] Basic Checks") {
```
I'd also add a test for `force_trivial` being set to `true`. It can be a copy of an existing test that uses a trivially destructible type, just to make sure it works correctly.

Ivorforce replied:
You mean check that it works in general, or that it works as `force_trivial`? For the latter I'd have no idea how! 
It's a good idea to check it's not broken though I agree, perhaps I'll just repurpose one of the test cases below. It's redundant enough as is.

lawnjelly reviewed (dismissed):
Looks fine to me, we can add some more functionality as required like `remove_unordered()` etc.

Ivorforce commented:
I moved `force_trivial` to the `resize` function only, in accordance to https://github.com/godotengine/godot/pull/104522.
In this PR, it was already used in just `resize`.

AThousandShips reviewed (dismissed):
The change to move `force_trivial` should wait for approval in #104522 and be discussed separately, so either this should wait for that PR to be approved or rejected, or this should be built with the original setup and then changed in that PR as well

Repiteo commented:
As ATS mentioned, could you revert the `force_trivial` changes so this can be assessed & integrated as-is? I just realized this was likely kept off the merge queue for that reason, and I'd like to keep this on our radar

Ivorforce commented:
It's still on my radar, just waiting for the next core meeting :)

Ivorforce commented:
I've updated the PR according to what was discussed in the core meeting:
- `resize` is removed. Instead, there are two new functions:
- `resize_initialized`, which _always_ initializes new elements, even if they are trivial types.
- `resize_uninitialized`, which _never_ initializes new elements. If the contained type is not trivially destructible, `FixedVector` statically complains for memory safety reasons.

I believe it's now ready to merge (from my side) :)

Ivorforce committed:
Add `FixedVector` template.
This is a high performance `Vector`-like object that can be used if the maximum number of objects is small and known, and the objects are needed only temporarily.

Ivorforce commented:
> One other idea (although not necessary) might be possible to check the alignment in the tests.

Good idea! Added.

Repiteo commented:
Thanks!

DeeJayLSP commented:
Wasn't sure where to put this, but from a quick look I found a few places `FixedVector` could potentially be a drop-in replacement for `Vector` and `LocalVector` (except you remove the `resize`s obviously).

These are either iterated in-function, or the functions that use them take either a pointer or `Span`.

- `BoneMapping::auto_mapping_process()` (`LocalVector<String>, 5`, one for each hand)
- `SpringBoneCollision3DGizmoPlugin::draw_capsule()` (`Vector3, 4`)
- `SpringBoneCollision3DGizmoPlugin::draw_plane()` (`Vector3, 7`)
- `BetsyCompressor::_compress()` (`float, 256 * 4`, possibly not safe)
- `SceneCacheInterface::process_simplify_path()` (`uint8_t, 6`)
- `TextServerAdvanced::_ensure_cache_for_size()` (`UChar, 255`)
- `RendererSceneCull::_light_instance_update_shadow()` (`Plane, 6`)

While compilation goes fine, these have shown immediate issues in a quick run:
- `RendererCanvasRenderRD::RendererCanvasRenderRD()` (`uint8_t, 6 * 2`, made `CanvasItem`s invisible)
- `RendererCompositorRD::initialize()` (`uint8_t, 6 * 2`, turned the whole engine into a big black screen)
